source/common/event/win32/signal_impl.cc

```diff
@@ -0,0 +1,40 @@
+#include "common/api/os_sys_calls_impl.h"
+#include "common/event/dispatcher_impl.h"
+#include "common/event/signal_impl.h"
+
+#include "event2/event.h"
+
+namespace Envoy {
+namespace Event {
+
+SignalEventImpl::SignalEventImpl(DispatcherImpl& dispatcher, signal_t signal_num, SignalCb cb)
+    : cb_(cb) {
+  auto handler_it = eventBridgeHandlersSingleton::get().find(signal_num);
+  if (handler_it != eventBridgeHandlersSingleton::get().end()) {
+    return;
+  }
+
+  os_fd_t socks[2];
+  Api::SysCallIntResult result =
+      Api::OsSysCallsSingleton::get().socketpair(AF_INET, SOCK_STREAM, IPPROTO_TCP, socks);
+  ASSERT(result.rc_ == 0);
+
+  read_handle_ = std::make_unique<Network::IoSocketHandleImpl>(socks[0], false, AF_INET);
+  result = read_handle_->setBlocking(false);
+  ASSERT(result.rc_ == 0);
+  auto write_handle = std::make_shared<Network::IoSocketHandleImpl>(socks[1], false, AF_INET);
+  result = write_handle->setBlocking(false);
+  ASSERT(result.rc_ == 0);
+
+  read_handle_->initializeFileEvent(
+      dispatcher,
+      [this](uint32_t events) -> void {
+        ASSERT(events == Event::FileReadyType::Read);
+        cb_();
+      },
+      Event::FileTriggerType::Level, Event::FileReadyType::Read);
+  eventBridgeHandlersSingleton::get().insert({signal_num, write_handle});
```

Comment:
What guarantees that this insert into the flat_hash_map ends up being thread safe?

Consider implementing this as an array instead of a map since the number of possible signal numbers is small.

Reply:
I assume the race condition that you refer to is someone hitting ctrl+c while the server main is adding new `signal_impl`. I transformed this into an array where each signal is assigned to a fixed position based on the signal id. Let me know what you think.